ci-studio-common-util/cmd/update.go

```diff
@@ -0,0 +1,163 @@
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/chef/ci-studio-common/lib"
+	"github.com/juju/fslock"
+	"github.com/mholt/archiver"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	updateCommand = &cobra.Command{
+		Use:   "update",
+		Short: "Update the ci-studio-common install",
+		Run:   maybeUpdateInstall,
+	}
+
+	updateOpts = struct {
+		phase  int
+		force  bool
+		suffix string
+	}{}
+
+	assetOnDisk = filepath.Join(lib.InstallDirParent, "ci-studio-common.tar.gz")
+
+	upgradeStatusFile = lib.SettingsPath("upgrade-in-progress")
+)
+
+func init() {
+	rootCmd.AddCommand(updateCommand)
+
+	updateCommand.Flags().IntVar(&updateOpts.phase, "phase", 1, "The phase of the update to complete")
+	updateCommand.Flags().BoolVar(&updateOpts.force, "force", false, "Perform the installation even if no updates are available")
+	updateCommand.Flags().StringVar(&updateOpts.suffix, "suffix", "rc", "The suffix to use when downloading the asset")
+}
+
+// When we run these phases, each phase triggers the next using cmd.Start(). We use Start() because it does
+// not wait for the command to finish, which is what we want for this sequence to work. We do it this way
+// because of limitations on Windows for binaries modifying themselves.
+func maybeUpdateInstall(cmd *cobra.Command, args []string) {
+	switch updateOpts.phase {
+	case 1:
+		lock := fslock.New(lib.LockPath("upgrade-ci-studio-common"))
+		lockErr := lock.TryLock()
+
+		if lockErr == nil {
+			performPhaseOne()
+		} else {
+			fmt.Println("ci-studio-common upgrade already in progress -- skipping")
+		}
+	case 2:
+		performPhaseTwo()
+	case 3:
+		performPhaseThree()
+	default:
+		log.Fatal("Unsupported phase")
+	}
+}
+
+// In phase one of the upgrade, we:
+//   1. check if an upgrade is required
+//   2. download the asset
+//   3. copy the existing directory into a backup directory
+//   4. begin phase two from the backup directory
+func performPhaseOne() {
+	assetURL := fmt.Sprintf("https://chef-cd-artifacts.s3-us-west-2.amazonaws.com/ci-studio-common/ci-studio-common-2.0.0-%s-%s.tar.gz", runtime.GOOS, updateOpts.suffix)
+
+	localEtag := lib.SettingWithDefault("etag", "none")
+	remoteEtag := lib.GetURLHeaderByKey(assetURL, "ETag")
+
+	if (remoteEtag != localEtag) || updateOpts.force {
+		fmt.Println("--> {1} Upgrade available")
+		err := os.Chdir(lib.InstallDirParent)
+		lib.Check(err)
+
+		// Mark that an upgrade is in progress
+		err = ioutil.WriteFile(upgradeStatusFile, []byte(""), 0644)
+		lib.Check(err)
+
+		// Save new etag to disk
+		err = ioutil.WriteFile(lib.SettingsPath("etag-new"), []byte(remoteEtag), 0644)
+		lib.Check(err)
+
+		// Download the asset to disk
+		fmt.Println("--> {1} Downloading latest release")
+		err = lib.DownloadFile(assetOnDisk, assetURL)
+		lib.Check(err)
+
+		// Try and backup the current installation
+		fmt.Println("--> {1} Release downloaded -- backuping up current installation")
```

Comment:
nit: since there are a number of commands, having a library to standardize the output displayed to users would be a nice to have.